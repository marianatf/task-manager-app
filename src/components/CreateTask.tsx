import React, { useState } from "react";
import { priorityColor } from "../utils/priorityColor";
import type { Priority } from "../types";

interface Props {
  addTask: (title: string, description: string, priority: Priority) => void;
}

const PRIORITIES: Array<Priority> = ["low", "medium", "high"];

export const CreateTask: React.FC<Props> = ({ addTask }) => {
  const [title, setTitle] = useState("");
  const [description, setDescription] = useState("");
  const [priority, setPriority] = useState<Priority>("low");
  const [isTitleFocused, setIsTitleFocused] = useState(false);

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault();
    addTask(title, description, priority);
    setTitle("");
    setDescription("");
    setPriority("low");
    setIsTitleFocused(false);
  };

  return (
    <form onSubmit={handleSubmit} className="flex flex-col gap-3 m-4">
      <input
        value={title}
        onChange={(e) => setTitle(e.target.value)}
        placeholder="Add a task"
        className="rounded-lg px-4 py-2 border border-gray-300 focus:outline-none focus:ring-2 focus:ring-black"
        onFocus={() => setIsTitleFocused(true)}
        onBlur={() => {
          if (!title) setIsTitleFocused(false);
        }}
      />
      {isTitleFocused && (
        <>
          <input
            value={description}
            onChange={(e) => setDescription(e.target.value)}
            placeholder="Add a description"
            className="rounded-lg px-4 py-2 border border-gray-300 focus:outline-none focus:ring-2 focus:ring-black"
          />
          <div className="flex gap-2">
            {PRIORITIES.map((priorityOption) => (
              <button
                key={priorityOption}
                type="button"
                className={`px-3 py-1 rounded-lg border border-gray-300 transition ${
                  priority === priorityOption
                    ? priorityColor(priorityOption)
                    : ""
                }`}
                onClick={() => setPriority(priorityOption)}
              >
                {priorityOption.charAt(0).toUpperCase() +
                  priorityOption.slice(1)}
              </button>
            ))}
          </div>
          <button
            type="submit"
            className="rounded-lg bg-black text-white px-4 py-2 hover:bg-black transition disabled:bg-gray-300 disabled:cursor-not-allowed"
            disabled={!title || !description}
          >
            Add
          </button>
        </>
      )}
    </form>
  );
};
