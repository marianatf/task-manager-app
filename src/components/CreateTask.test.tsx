import { CreateTask } from "./CreateTask";
import { render, fireEvent, screen } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";

describe("CreateTask", () => {
  const mockAddTask = vi.fn();

  it("should match snapshot", () => {
    const { asFragment } = render(<CreateTask addTask={mockAddTask} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should call addTask", () => {
    render(<CreateTask addTask={mockAddTask} />);

    const titleInput = screen.getByPlaceholderText("Add a task");
    fireEvent.focus(titleInput);
    fireEvent.change(titleInput, { target: { value: "task" } });

    const descriptionInput = screen.getByPlaceholderText("Add a description");
    fireEvent.change(descriptionInput, { target: { value: "description" } });

    fireEvent.click(screen.getByText("High"));
    fireEvent.click(screen.getByRole("button", { name: /add/i }));

    expect(mockAddTask).toHaveBeenCalledWith("task", "description", "high");
  });

  it("should show description and priority when title is focused", () => {
    render(<CreateTask addTask={mockAddTask} />);

    const titleInput = screen.getByPlaceholderText("Add a task");
    fireEvent.focus(titleInput);
    const descriptionInput = screen.getByPlaceholderText("Add a description");
    expect(descriptionInput).toBeInTheDocument();
    expect(screen.getByText("Low")).toBeInTheDocument();
    expect(screen.getByText("Medium")).toBeInTheDocument();
    expect(screen.getByText("High")).toBeInTheDocument();
  });

  it("should update title and description", () => {
    render(<CreateTask addTask={mockAddTask} />);

    const titleInput = screen.getByPlaceholderText("Add a task");
    fireEvent.focus(titleInput);
    fireEvent.change(titleInput, { target: { value: "task" } });
    expect(titleInput).toHaveValue("task");

    const descriptionInput = screen.getByPlaceholderText("Add a description");
    fireEvent.change(descriptionInput, { target: { value: "description" } });
    expect(descriptionInput).toHaveValue("description");
  });

  it("should change priority when clicked", () => {
    render(<CreateTask addTask={mockAddTask} />);

    const titleInput = screen.getByPlaceholderText("Add a task");
    fireEvent.focus(titleInput);
    const highPriorityBtn = screen.getByText("High");
    fireEvent.click(highPriorityBtn);
    expect(highPriorityBtn.className).toMatch(
      "px-3 py-1 rounded-lg border border-gray-300 transition bg-red-200"
    );
  });

  it("should disable submit button correctly", () => {
    render(<CreateTask addTask={mockAddTask} />);

    const titleInput = screen.getByPlaceholderText("Add a task");
    fireEvent.focus(titleInput);
    const addBtn = screen.getByRole("button", { name: /add/i });
    expect(addBtn).toBeDisabled();

    fireEvent.change(titleInput, { target: { value: "task" } });
    expect(addBtn).toBeDisabled();

    const descriptionInput = screen.getByPlaceholderText("Add a description");
    fireEvent.change(descriptionInput, { target: { value: "description" } });
    expect(addBtn).not.toBeDisabled();
  });
});
