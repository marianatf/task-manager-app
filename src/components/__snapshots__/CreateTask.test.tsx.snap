// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateTask > should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="flex flex-col gap-3 m-4"
  >
    <input
      class="rounded-lg px-4 py-2 border border-gray-300 focus:outline-none focus:ring-2 focus:ring-black"
      placeholder="Add a task"
      value=""
    />
  </form>
</DocumentFragment>
`;

exports[`Header > should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="flex flex-col gap-3 m-4"
  >
    <input
      class="rounded-lg px-4 py-2 border border-gray-300 focus:outline-none focus:ring-2 focus:ring-black"
      placeholder="Add a task"
      value=""
    />
  </form>
</DocumentFragment>
`;
